streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java

```diff
@@ -0,0 +1,228 @@
+                    // else if topic exists with more #.partitions than needed, delete in order to re-create it
+                    try {
+                        deleteTopic(topic);
+                    } catch (ZkNodeExistsException e) {
+                        // ignore and continue
+                    }
+                } else if (topicMetadata.size() < numPartitions) {
+                    // else if topic exists with less #.partitions than needed, add partitions
+                    try {
+                        addPartitions(topic, numPartitions - topicMetadata.size(), replicationFactor, topicMetadata);
+                    } catch (ZkNoNodeException e) {
+                        // ignore and continue
+                    }
+                }
+
+                topicNotReady = false;
+            }
+        }
+    }
+
+    private List<Integer> getBrokers() {
+        List<Integer> brokers = new ArrayList<>();
+        for (String broker: zkClient.getChildren(ZK_BROKER_PATH)) {
+            brokers.add(Integer.parseInt(broker));
+        }
+        Collections.sort(brokers);
+
+        log.debug("Read brokers {} from ZK in partition assignor.", brokers);
+
+        return brokers;
+    }
+
+    @SuppressWarnings("unchecked")
+    public Map<Integer, List<Integer>> getTopicMetadata(String topic) {
+        String data = zkClient.readData(ZK_TOPIC_PATH + "/" + topic, true);
+
+        if (data == null) return null;
+
+        try {
+            ObjectMapper mapper = new ObjectMapper();
+
+            Map<String, Object> dataMap = mapper.readValue(data, new TypeReference<Map<String, Object>>() {
+
+            });
+
+            Map<Integer, List<Integer>> partitions = (Map<Integer, List<Integer>>) dataMap.get("partitions");
+
+            log.debug("Read partitions {} for topic {} from ZK in partition assignor.", partitions, topic);
+
+            return partitions;
+        } catch (IOException e) {
+            throw new StreamsException("Error while reading topic metadata from ZK for internal topic " + topic, e);
+        }
+    }
+
+    public void createTopic(String topic, int numPartitions, int replicationFactor) throws ZkNodeExistsException {
+        log.debug("Creating topic {} with {} partitions from ZK in partition assignor.", topic, numPartitions);
+
+        List<Integer> brokers = getBrokers();
+        int numBrokers = brokers.size();
+        if (numBrokers < replicationFactor) {
+            log.warn("Not enough brokers found. The replication factor is reduced from " + replicationFactor + " to " +  numBrokers);
+            replicationFactor = numBrokers;
+        }
+
+        Map<Integer, List<Integer>> assignment = new HashMap<>();
+
+        for (int i = 0; i < numPartitions; i++) {
+            ArrayList<Integer> brokerList = new ArrayList<>();
+            for (int r = 0; r < replicationFactor; r++) {
+                int shift = r * numBrokers / replicationFactor;
```

Comment:
This shifting mechanism is inconsistent with Kafka's create / alter topic tools, any reason why you want to use this manner?